File: problems/LeetCode/tree/levelOrder.py
def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:

		if not root: return []

		ans = []
		queue = []
		queue = deque([(0, root)])
		dic = defaultdict(list)

		while len(queue):
				depth, node = queue.popleft()
				dic[depth].append(node.val)

				if node.left:
						queue.append([depth + 1, node.left])
				if node.right:
						queue.append([depth + 1, node.right])

		for value in dic:
				ans.append(dic[value])
		
		return ans


def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
		if not root: return []

		ans = []
		q = deque([root])

		while q:

				vals = []

				for _ in range(len(q)):
						node = q.popleft()
						vals.append(node.val)
						if node.left: q.append(node.left)
						if node.right: q.append(node.right)
				
				ans.append(vals)

		return ans

File: problems/LeetCode/TopInterview150/levelOrder.ts
function levelOrder(root: TreeNode | null): number[][] {
  if (!root) return [];

  let queue = [root];
  let ans = [];

  while (queue.length) {
    let len = queue.length;
    const levelNodes = [];

    for (let i = 0; i < len; i++) {
      const node = queue[i];
      levelNodes.push(node.val);
      if (node.left) queue.push(node.left);
      if (node.right) queue.push(node.right);
    }

    ans.push(levelNodes);
    queue = queue.slice(len);
  }

  return ans;
}

function levelOrder(root: TreeNode | null): number[][] {
  if (!root) return [];

  let queue = [root];
  let ans = [];
  let head = 0;

  while (head < queue.length) {
    const levelEnd = queue.length;
    const level: number[] = [];
    while (head < levelEnd) {
      level.push(queue[head].val);
      if (queue[head].left) queue.push(queue[head].left);
      if (queue[head].right) queue.push(queue[head].right);
      head++;
    }
    ans.push(level);
  }

  return ans;
}

// Time complexity: O(n)
// Space complexity: O(n)
